github/codeowners.py

```diff
@@ -81,73 +111,173 @@ def filter_codeowners_entries(
         yield from line.split()[1:]
 
 
-def _first(
-    iterable: typing.Iterable,
-):
+def resolve_team_members(
+    team_name: str,
+    org: github3.orgs.Organization,
+    absent_ok: bool=False,
+) -> typing.Generator[github3.users.ShortUser, None, None]:
+    '''
+    Returns generator yielding members from team.
+    Absent teams passed silently with `absent_ok`.
+    Subteams are resolved as well.
+    '''
     try:
-        return next(iterable)
-    except StopIteration:
-        return None
+        team = org.team_by_name(team_name)
+    except github3.exceptions.NotFoundError:
+        if absent_ok:
+            return
+        raise
+    # resolves subteams as well
+    yield from team.members()
 
 
-def determine_email_address(
-    github_user_name: str,
-    github_api: GitHub,
-) -> typing.Optional[str]:
-    not_none(github_user_name)
+def mail_address_for_user(
+    user_name: str,
+    gh_api: GitHub | GitHubEnterprise,
+    absent_ok: bool=True,
+) -> str | None:
+    '''
+    Returns email associated with github user for username.
+    `None` returned if user not found and `absent_ok`.
+    Users without associated email return `None`.
+    '''
     try:
-        user = github_api.user(github_user_name)
+        user = gh_api.user(user_name)
     except NotFoundError:
-        logger.warning(f'failed to lookup {github_user_name=} {github_api._github_url=}')
-        return None
+        if absent_ok:
+            logger.warning(f'failed to lookup {user_name=} {gh_api._github_url=}')
+            return None
+        raise
 
     return user.email
 
 
-def resolve_team_members(
-    github_team_name: str,
-    github_api: GitHub,
-) -> typing.Union[typing.Generator[str, None, None], list]:
-    not_none(github_team_name)
-    org_name, team_name = github_team_name.split('/') # always of form 'org/name'
-    organisation = github_api.organization(org_name)
-    # unfortunately, we have to look-up the team (no api to retrieve it by name)
-    team_or_none = _first(filter(lambda team: team.slug == team_name, organisation.teams()))
-    if not team_or_none:
-        logger.warning('failed to lookup team {t}'.format(t=team_name))
-        return []
-    for member in map(github_api.user, team_or_none.members()):
-        if member.email:
-            yield member.email
-        else:
-            logger.warning(f'no email found for GitHub user {member}')
+def user_for_mail_address(
+    mail_address: str,
+    gh_api: GitHub | GitHubEnterprise,
+) -> github3.users.ShortUser | None:
+    '''
+    Find user for email address.
+    Exactly one user for `None` will be returned.
```

Comment:
> Exactly one user for `None` will be returned.

1. I think this should actually read _or_ rather than _for_?
2. this just repeats the typehint -> no added value -> can be removed